### ISSUE TITLE
CatBoostClassifier creates inaccurate confusion matrices when class labels are reduced and the predictions are passed through scikit-learn 

### ISSUE THREAD

**Issue by patentedata** [catboost/catboost#1127]
Inaccurate scikit-learn confusion matrix with removed classes:
```
from pandas import DataFrame
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool
from catboost.utils import get_confusion_matrix
from sklearn.metrics import confusion_matrix

example = {'feature_1': [1.0,1.0,2.0,7.0,2.0,9.0,7.0, 9.0,2.0,2.0,9.0,9.0,7.0,7.0,1.0,1.0],
           'feature_2': [250,800,200,300,500,180,100,300,500,120,50,100,550,400,850,700]
          }

df = DataFrame(example, columns= ['feature_1', 'feature_2'])

print("All 4 classes in the desired label 'feature_1':")
print(df['feature_1'].value_counts())
print()

df = df[(df.feature_1 != 7.0) & (df.feature_1 != 9.0)]

print("The 2 remaining classes in 'feature_1' after the first 2 are removed:")
print(df['feature_1'].value_counts())
print()

train, test = train_test_split(df, 
                              train_size=0.80, 
                              test_size=0.20, 
                              stratify=df['feature_1'], 
                              random_state=42)

label      = 'feature_1'
net_feats  = train.drop(columns=[label])
features   = net_feats.columns.tolist()

X_train  = train[features]
y_train  = train[label]
X_test   = test[features]
y_test   = test[label]

model = CatBoostClassifier(
                           learning_rate=0.03, 
                           random_seed=42,
                          )

model.fit(X_train,
          y_train,
          eval_set=(X_test, y_test),
            verbose=False)

y_preds      = model.predict(X_train)
y_preds_test = model.predict(X_test)

train_pool = Pool(X_train, y_train)
test_pool  = Pool(X_test, y_test)

print('Since 2 out of 4 classes are included in the label, a 2x2')
print('confusion matrix is exptected after fitting the model.')
print('However, a 3x3 matrix is produced with scikit-learn:')
print()
print('CatBoost train_pool confusion matrix:')
print(get_confusion_matrix(model, train_pool, thread_count=-1))
print()
print('scikit-learn y_train confusion matrix:')
print(confusion_matrix(y_train, y_preds))
print()
print('CatBoost test_pool confusion matrix:')
print(get_confusion_matrix(model, test_pool, thread_count=-1))
print()
print('scikit-learn y_test confusion matrix:')
print(confusion_matrix(y_test, y_preds_test))
```

Accurate scikit-learn confusion matrix with all the classes:
```
from pandas import DataFrame
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool
from catboost.utils import get_confusion_matrix
from sklearn.metrics import confusion_matrix

example = {'feature_1': [1.0,1.0,2.0,7.0,2.0,9.0,7.0, 9.0,2.0,2.0,9.0,9.0,7.0,7.0,1.0,1.0],
           'feature_2': [250,800,200,300,500,180,100,300,500,120,50,100,550,400,850,700]
          }

df = DataFrame(example, columns= ['feature_1', 'feature_2'])

print("All 4 classes in the desired label 'feature_1':")
print(df['feature_1'].value_counts())
print()

# df = df[(df.feature_1 != 7.0) & (df.feature_1 != 9.0)]

# print("The 2 remaining classes in 'feature_1' after the first 2 are removed:")
# print(df['feature_1'].value_counts())
# print()

train, test = train_test_split(df, 
                              train_size=0.80, 
                              test_size=0.20, 
                              stratify=df['feature_1'], 
                              random_state=42)

label      = 'feature_1'
net_feats  = train.drop(columns=[label])
features   = net_feats.columns.tolist()

X_train  = train[features]
y_train  = train[label]
X_test   = test[features]
y_test   = test[label]

model = CatBoostClassifier(
                           learning_rate=0.03, 
                           random_seed=42,
                          )

model.fit(X_train,
          y_train,
          eval_set=(X_test, y_test),
            verbose=False)

y_preds      = model.predict(X_train)
y_preds_test = model.predict(X_test)

train_pool = Pool(X_train, y_train)
test_pool  = Pool(X_test, y_test)

print("Since all 4 classes are in the label, a 4x4 confusion matrix")
print("is exptected after fitting the model. Unlike the instance where")
print("2 out of 4 classes were dropped and the model produced an")
print("inaccurate 3x3 matrix with the 2 remaining classes, an accurate ")
print("4x4 confusion matrix was produced with scikit-learn:")
print()
print('CatBoost train_pool confusion matrix:')
print(get_confusion_matrix(model, train_pool, thread_count=-1))
print()
print('scikit-learn y_train confusion matrix:')
print(confusion_matrix(y_train, y_preds))
print()
print('CatBoost test_pool confusion matrix:')
print(get_confusion_matrix(model, test_pool, thread_count=-1))
print()
print('scikit-learn y_test confusion matrix:')
print(confusion_matrix(y_test, y_preds_test))
print()
print('If the issue of erroneous confusion matrices was consistent')
print('with removal and non-removal class instances, we would see')
print('5x5 matrices produced above.')
```

Accurate scikit-learn confusion matrix with removed classes and remapping of the remaining classes:
```
from pandas import DataFrame
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool
from catboost.utils import get_confusion_matrix
from sklearn.metrics import confusion_matrix

example = {'feature_1': [1.0,1.0,2.0,7.0,2.0,9.0,7.0, 9.0,2.0,2.0,9.0,9.0,7.0,7.0,1.0,1.0],
           'feature_2': [250,800,200,300,500,180,100,300,500,120,50,100,550,400,850,700]
          }

df = DataFrame(example, columns= ['feature_1', 'feature_2'])

print("All 4 classes in the desired label 'feature_1':")
print(df['feature_1'].value_counts())
print()

df = df[(df.feature_1 != 7.0) & (df.feature_1 != 9.0)]
df['feature_1'] = df['feature_1'].map({1.0: 0.0, 2.0: 1.0})

print("The 2 remaining classes in 'feature_1' after the first 2 are removed")
print("and the other 2 are remapped:")
print(df['feature_1'].value_counts())
print()

train, test = train_test_split(df, 
                              train_size=0.80, 
                              test_size=0.20, 
                              stratify=df['feature_1'], 
                              random_state=42)

label      = 'feature_1'
net_feats  = train.drop(columns=[label])
features   = net_feats.columns.tolist()

X_train  = train[features]
y_train  = train[label]
X_test   = test[features]
y_test   = test[label]

model = CatBoostClassifier(
                           learning_rate=0.03, 
                           random_seed=42,
                          )

model.fit(X_train,
          y_train,
          eval_set=(X_test, y_test),
            verbose=False)

y_preds      = model.predict(X_train)
y_preds_test = model.predict(X_test)

train_pool = Pool(X_train, y_train)
test_pool  = Pool(X_test, y_test)

print('Since there are only 2 classes in the label, a 2x2 confusion')
print('matrix is exptected after fitting the model. Unlike the instance')
print('where removing 2 out of 4 classes produced an inaccurate 3x3 matrix,')
print('remapping the remaining 2 classes after the class removal process')
print('created an accurate 2x2 matrix with both CatBoost and scikit-learn:')
print()
print('CatBoost train_pool confusion matrix:')
print(get_confusion_matrix(model, train_pool, thread_count=-1))
print()
print('scikit-learn y_train confusion matrix:')
print(confusion_matrix(y_train, y_preds))
print()
print('CatBoost test_pool confusion matrix:')
print(get_confusion_matrix(model, test_pool, thread_count=-1))
print()
print('scikit-learn y_test confusion matrix:')
print(confusion_matrix(y_test, y_preds_test))
```

If this cross-library issue cannot be addressed, can better confusion matrix visualizations be incorporated into CatBoost? Something like [this](https://i.stack.imgur.com/NuOUD.png) would be extremely helpful.

**Comment by patentedata** [comment]
@annaveronika

**Comment by annaveronika** [comment]
Thanks for the issue, we'll look into this.

**Comment by LyzhinIvan** [comment]
Hi, @patentedata !

What versions of catboost and sklearn are you using? I have run your scripts with catboost 0.21 and sklearn 0.21.2. Everything works correctly.

**Comment by annaveronika** [comment]
Closing because of inactivity. Hope the issue is already resolved.
